openerp/models.py

```diff
@@ -4249,16 +4251,19 @@ def _create(self, cr, user, vals, context=None):
         # check Python constraints
         recs._validate_fields(vals)
 
-        if recs.env.recompute and context.get('recompute', True):
-            result += self._store_get_values(cr, user, [id_new],
+        result += self._store_get_values(cr, user, [id_new],
                 list(set(vals.keys() + self._inherits.values())),
                 context)
+        recs.env.extend_recompute_list(result)
+
+        if recs.env.recompute and context.get('recompute', True):
             result.sort()
             done = []
-            for order, model_name, ids, fields2 in result:
+            for order, model_name, ids, fields2 in recs.env.get_recompute_list():
                 if not (model_name, ids, fields2) in done:
                     self.pool[model_name]._store_set_values(cr, user, ids, fields2, context)
                     done.append((model_name, ids, fields2))
+            recs.env.clear_recompute_list()
```

Comment:
I see a potential issue here: while you execute method `_store_set_values`, you may trigger more recomputations that will be added to `recompute_list`, which has not been cleared yet. This may cause two issues:
 - the functions to recompute will be recomputed once again in the recursive call to `create`/`write`;
 - the iteration on `recompute_list` may fail because new elements have been added.

In view of this problem, I suggest to combine the methods `get_recompute_list` and `clear_recompute_list`. The API would become something like:

    class Environment:

        def extend_recompute_old(self, items):
            self.all.recompute_old += items

        def get_recompute_old(self):
            result = self.all.recompute_old
            self.all.recompute_old = []
            return result

In the code that processes the recomputation,
 - remove the call to `clear_recompute_list`;
 - sort the list you get from `get_recompute_list` before iterating over it.